### PR TITLE
Refactor CLI for new API

### LINKED ISSUE #103: Refactor CLI to use new api Storage

### LINKED ISSUE #164: rename cli from `io` to new name
in the process of de-amazeeiofying Lagoon we also need to find a new name for the CLI (currently called `io`).

My own brainstorming came up with these ideas:
- `lagoon` - well, lame :D plus also very long
- `goon` - a bit used by Defcon an it's Goons already
- `lagu` - my most prefeered name so far, short, has similarities to lagoon. Not used by anything really

what do you think?

### PR DESCRIPTION
**Pull request by karlhorky**
Based on changes described in #35 and #164

Closes #103 
Closes #164

### TODO

- [x] `customer`
- [x] `project`
- [x] `projects`
- [x] `project create`
  - [x] Working command with mutation
  - [x] Tests
  - [x] Examples
- [x] Add tests
- [x] Fix new Flow errors
- [x] Update all occurrences of `io` to `lagu`
  - [x] Update name in docs
  - [x] Update auth token filename
- [x] Move environment variables from `runCli.sh` to `.lagoon.yml` and delete `runCli.sh`
  - [x] get API URL from `api:` in  [`.lagoon.yml`](https://github.com/amazeeio/drupal-example/blob/1ce16c0641c127633bb68f1433df68f1bb855a78/.lagoon.yml#L4-L5)
  - [x] get SSH_HOST and SSH_PORT from `endpoint:` in  [`.lagoon.yml`](https://github.com/amazeeio/drupal-example/blob/1ce16c0641c127633bb68f1433df68f1bb855a78/.lagoon.yml#L4-L5)
  - [x] Pass (or curry in cli.js) the needed config parameters to `sshConnect` and `sshExec`
  - [x] Pass (or curry in cli.js) the needed config parameters to `runGQLQuery`
  - [x] Remove `process.env` and default value for `endpoint` parameters from `query.js`, etc
  - [x] Make it possible to run the CLI in some sort of development mode (`babel --watch` + symlink?)
- [x] During Project create it asks for an URL and validates it, which is great, but unfortunately `git@github.com:amazeeio/drupal-example.git` is not seen as a valid URL, which is though a git ssh URL
- [x] Define what node version we want to support (>=8) and improve the transpilation with `@babel/env` ([like](https://github.com/jaydenseric/apollo-upload-client/blob/master/.babelrc.js) [this](https://github.com/jaydenseric/apollo-upload-client/blob/master/package.json)?)
- [x] Make it less annoying to run the `login` and `logout` commands in dev mode (`process.env` with something like [transform-inline-environment-variables](https://babeljs.io/docs/plugins/transform-inline-environment-variables/#via-node-api)?)

### TODO for other PRs

- ~~Make CLI installable~~ already done
- Maybe for followup PRs: add e2e tests with [`inquirer-test`](https://github.com/ewnd9/inquirer-test)
- Update CLI documentation in readme
- Move logging to separate file (avoid having to pass `clog` and `cerr` everywhere), fix tests (change mocks, [add Jest setup for disabling logging to console during unit test runs](https://stackoverflow.com/questions/38363292/disable-winston-logging-when-running-unit-tests))

### `customer`

```
# Without comment, with deploy private key
../runCli.sh customer
yarn run v1.3.2
$ node -r babel-core/register src/cli.js customer
Customer information for project 'ci-bitbucket'
╔════════════════════╤═════════════════════╗
║ Name               │ ci-customer         ║
╟────────────────────┼─────────────────────╢
║ Comment            │                     ║
╟────────────────────┼─────────────────────╢
║ Deploy Private Key │ √                   ║
╟────────────────────┼─────────────────────╢
║ SSH Keys           │ ci-customer-sshkey  ║
╟────────────────────┼─────────────────────╢
║ Created            │ 2018-01-15 11:09:35 ║
╚════════════════════╧═════════════════════╝

# With comment, without deploy private key
../runCli.sh customer
yarn run v1.3.2
$ node -r babel-core/register src/cli.js customer
Customer information for project 'credentialstest-project1'
╔════════════════════╤═════════════════════════════════════════════════════════════╗
║ Name               │ credentialtest-customer1                                    ║
╟────────────────────┼─────────────────────────────────────────────────────────────╢
║ Comment            │ used to test the cli                                        ║
╟────────────────────┼─────────────────────────────────────────────────────────────╢
║ Deploy Private Key │ ✗                                                           ║
╟────────────────────┼─────────────────────────────────────────────────────────────╢
║ SSH Keys           │ credentialtestbothcustomeraccess, credentialtestcrossaccess ║
╟────────────────────┼─────────────────────────────────────────────────────────────╢
║ Created            │ 2018-01-15 11:09:35                                         ║
╚════════════════════╧═════════════════════════════════════════════════════════════╝
```

### `projects`

```
../runCli.sh projects
yarn run v1.3.2
$ node -r babel-core/register src/cli.js projects
╔═════════════════════════╤═══════════════════════════════════════════════╤══════════╤═══════════════╤═════════════════════╗
║ Project                 │ Git URL                                       │ Branches │ Pull Requests │ Created             ║
╟─────────────────────────┼───────────────────────────────────────────────┼──────────┼───────────────┼─────────────────────╢
║ ci-bitbucket            │ ssh://git@10.0.2.2:32768/git/bitbucket.git    │ true     │ null          │ 2017-12-06 17:22:13 ║
╟─────────────────────────┼───────────────────────────────────────────────┼──────────┼───────────────┼─────────────────────╢
║ ci-drupal               │ ssh://git@10.0.2.2:32768/git/drupal.git       │ true     │ null          │ 2017-12-06 17:22:13 ║
╟─────────────────────────┼───────────────────────────────────────────────┼──────────┼───────────────┼─────────────────────╢
║ ci-github               │ ssh://git@10.0.2.2:32768/git/github.git       │ true     │ null          │ 2017-12-06 17:22:13 ║
╟─────────────────────────┼───────────────────────────────────────────────┼──────────┼───────────────┼─────────────────────╢
║ ci-gitlab               │ ssh://git@10.0.2.2:32768/git/gitlab.git       │ true     │ null          │ 2017-12-06 17:22:13 ║
╟─────────────────────────┼───────────────────────────────────────────────┼──────────┼───────────────┼─────────────────────╢
║ ci-multiproject1        │ ssh://git@10.0.2.2:32768/git/multiproject.git │ true     │ null          │ 2017-12-06 17:22:13 ║
╟─────────────────────────┼───────────────────────────────────────────────┼──────────┼───────────────┼─────────────────────╢
║ ci-multiproject2        │ ssh://git@10.0.2.2:32768/git/multiproject.git │ true     │ null          │ 2017-12-06 17:22:13 ║
╟─────────────────────────┼───────────────────────────────────────────────┼──────────┼───────────────┼─────────────────────╢
║ ci-nginx                │ ssh://git@10.0.2.2:32768/git/nginx.git        │ true     │ null          │ 2017-12-06 17:22:13 ║
╟─────────────────────────┼───────────────────────────────────────────────┼──────────┼───────────────┼─────────────────────╢
║ ci-node                 │ ssh://git@10.0.2.2:32768/git/node.git         │ true     │ null          │ 2017-12-06 17:22:13 ║
╟─────────────────────────┼───────────────────────────────────────────────┼──────────┼───────────────┼─────────────────────╢
║ ci-rest                 │ ssh://git@10.0.2.2:32768/git/rest.git         │ true     │ null          │ 2017-12-06 17:22:13 ║
╟─────────────────────────┼───────────────────────────────────────────────┼──────────┼───────────────┼─────────────────────╢
║ credentialstest-project │ ssh://git@10.0.2.2:32768/git/github.git       │ true     │ null          │ 2017-12-06 17:22:13 ║
╟─────────────────────────┼───────────────────────────────────────────────┼──────────┼───────────────┼─────────────────────╢
║ lagoon                  │ git@github.com:amazeeio/lagoon.git            │ true     │ null          │ 2017-12-06 17:22:13 ║
╚═════════════════════════╧═══════════════════════════════════════════════╧══════════╧═══════════════╧═════════════════════╝
```

### `project`

```
../runCli.sh project ci-bitbucket
yarn run v1.3.2
$ node -r babel-core/register src/cli.js project ci-bitbucket
Project details for 'ci-bitbucket':
╔═══════════════════════╤════════════════════════════════════════════╗
║ Project Name          │ ci-bitbucket                               ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Customer              │ ci-customer                                ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Git URL               │ ssh://git@10.0.2.2:32768/git/bitbucket.git ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Active Systems Deploy │ lagoon_openshiftBuildDeploy                ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Active Systems Remove │ lagoon_openshiftRemove                     ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Branches              │ true                                       ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Pull Requests         │ null                                       ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Openshift             │ ci-local                                   ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Created               │ 2017-12-06 17:22:13                        ║
╚═══════════════════════╧════════════════════════════════════════════╝
```

### `project create`

```
# With only single authorized customer / openshift
../runCli.sh project create
yarn run v1.3.2
$ node -r babel-core/register src/cli.js project create
! Using only authorized customer "credentialtest-customer1"
? Project name: much project
? Git URL: ssh://git@192.168.99.1:2222/git/rest.git
! Using only authorized openshift "credentialtest-openshift"
? Active system for task "deploy": lagoon_openshiftBuildDeploy
? Active system for task "remove": lagoon_openshiftRemove
? Deploy branches: true
? Pull requests: 
? Production environment: 
Project "much project" created successfully:
╔═══════════════════════╤══════════════════════════════════════════╗
║ Project Name          │ much project                             ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Customer              │ credentialtest-customer1                 ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Git URL               │ ssh://git@192.168.99.1:2222/git/rest.git ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Active Systems Deploy │ lagoon_openshiftBuildDeploy              ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Active Systems Remove │ lagoon_openshiftRemove                   ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Branches              │ true                                     ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Pull Requests         │ null                                     ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Openshift             │ credentialtest-openshift                 ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Created               │ 2018-01-31 17:30:16                      ║
╚═══════════════════════╧══════════════════════════════════════════╝

✨  Done in 30.75s.

# With multiple authorized customers / openshifts
../runCli.sh project create
yarn run v1.3.2
$ node -r babel-core/register src/cli.js project create
? Customer: (Use arrow keys)
❯ credentialtest-customer1 
  credentialtest-customer2 
  ci-customer 
  kickstart-customer 
? Project name: so projectular
? Git URL: ssh://git@192.168.99.1:2222/git/rest.git
? Openshift: 
  credentialtest-openshift 
❯ ci-local 
  kickstart-openshift
? Active system for task "deploy": lagoon_openshiftBuildDeploy
? Active system for task "remove": lagoon_openshiftRemove
? Deploy branches: true
? Pull requests: 
? Production environment: 
Project "so projectular" created successfully:
╔═══════════════════════╤══════════════════════════════════════════╗
║ Project Name          │ so projectular                           ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Customer              │ credentialtest-customer1                 ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Git URL               │ ssh://git@192.168.99.1:2222/git/rest.git ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Active Systems Deploy │ lagoon_openshiftBuildDeploy              ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Active Systems Remove │ lagoon_openshiftRemove                   ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Branches              │ true                                     ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Pull Requests         │ null                                     ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Openshift             │ ci-local                                 ║
╟───────────────────────┼──────────────────────────────────────────╢
║ Created               │ 2018-01-31 17:33:43                      ║
╚═══════════════════════╧══════════════════════════════════════════╝

✨  Done in 71.68s.
```

### .lagoon.yml API / SSH Configuration

```
api: http://localhost:3000/graphql
endpoint: ssh://lagoon@localhost:2020
```

### CODE REVIEW

**Comment by karlhorky** [comment]
~~Once this is merged, we can open a PR from my branch which updates Flow and fixes the resulting errors:~~

~~https://github.com/amazeeio/lagoon/compare/103-refactor-cli-for-new-api...update-flow-and-fix-errors?expand=1~~

---

**Edit:** We've merged `update-flow-and-fix-errors` into this branch beforehand: #294

**Review by ryyppy — COMMENTED** [review]
Please make sure to rename `amazee.io` occurrences to `lagoon`... i made some line comments, but maybe there are some slip throughs

File: cli/README.md [hunk 1]
```diff
@@ -76,18 +76,18 @@ This will generate the following file:
...
     before_deploy: []
     after_deploy: []
 ```
 
-### `io login`
+### `lagoon login`
 
 ```text
-$ io login --help
-io login - Authenticate with amazee.io via an SSH key
```
[the remark this answers is not in the document]
Comment (karlhorky) [reply cli/README.md]: Done in all occurrences, thanks!

**Comment by karlhorky** [comment]
**Edit:** Fixed in latest merge from master

Hm, tests are failing right now with the Mongo build. Probably not caused by this PR though.

```
docker build --quiet --build-arg IMAGE_REPO=jenkinslagoonpr11162 -t jenkinslagoonpr11162/mongo -f images/mongo/Dockerfile images/mongo

Sending build context to Docker daemon   2.56kB


Step 1/15 : ARG IMAGE_REPO

Step 2/15 : FROM ${IMAGE_REPO:-lagoon}/commons as commons

 ---> 84110abcab6e

Step 3/15 : FROM alpine:edge

 ---> 5c4fa780951b

Step 4/15 : MAINTAINER amazee.io

 ---> Using cache

 ---> 6968e8f6d2f0

Step 5/15 : ENV LAGOON=mongo

 ---> Using cache

 ---> d991b996a6cf

Step 6/15 : COPY --from=commons /lagoon /lagoon

 ---> Using cache

 ---> ad4b36ee550c

Step 7/15 : COPY --from=commons /bin/fix-permissions /bin/ep /bin/docker-sleep /bin/

 ---> Using cache

 ---> 47bbe426751d

Step 8/15 : COPY --from=commons /home /home

 ---> Using cache

 ---> 912ab65995b2

Step 9/15 : ENV TMPDIR=/tmp TMP=/tmp HOME=/home ENV=/home/.bashrc BASH_ENV=/home/.bashrc

 ---> Using cache

 ---> 31c9d6fb0b24

Step 10/15 : RUN apk --no-cache add mongodb

 ---> Running in c1edab623e7a

fetch http://dl-cdn.alpinelinux.org/alpine/edge/main/x86_64/APKINDEX.tar.gz

fetch http://dl-cdn.alpinelinux.org/alpine/edge/community/x86_64/APKINDEX.tar.gz

ERROR: unsatisfiable constraints:

  so:libyaml-cpp.so.0.5 (missing):

    required by:

                 mongodb-3.6.3-r0[so:libyaml-cpp.so.0.5]

                 mongodb-3.6.3-r0[so:libyaml-cpp.so.0.5]

                 mongodb-3.6.3-r0[so:libyaml-cpp.so.0.5]

                 mongodb-3.6.3-r0[so:libyaml-cpp.so.0.5]

                 mongodb-3.6.3-r0[so:libyaml-cpp.so.0.5]

                 mongodb-3.6.3-r0[so:libyaml-cpp.so.0.5]

The command '/bin/sh -c apk --no-cache add mongodb' returned a non-zero code: 2
```

**Comment by karlhorky** [comment]
I've updated the `babel` packages to the v7 beta versions and configured `@babel/env` to transpile less. Nice filesize reductions (see below).

I also made the formatting / linting config consistent across the workspaces that use it with a common shared package.

**Before:**

```
tree -h
.
├── [ 128]  __tests__
│   ├── [2.8K]  printErrors.test.js
│   └── [6.0K]  query.test.js
├── [3.6K]  cli.js
├── [ 352]  commands
│   ├── [ 192]  __tests__
│   │   ├── [3.4K]  customer.test.js
│   │   ├── [3.5K]  init.test.js
│   │   ├── [3.6K]  project.test.js
│   │   └── [3.6K]  projects.test.js
│   ├── [5.1K]  customer.js
│   ├── [  13]  index.js
│   ├── [6.4K]  init.js
│   ├── [3.0K]  login.js
│   ├── [3.1K]  logout.js
│   ├── [5.1K]  project.js
│   ├── [ 128]  projectCommands
│   │   ├── [  96]  __tests__
│   │   │   └── [ 11K]  project-create.test.js
│   │   └── [ 20K]  create.js
│   └── [3.7K]  projects.js
├── [ 128]  config
│   ├── [  96]  __tests__
│   │   └── [2.9K]  config.test.js
│   └── [1.6K]  index.js
├── [ 575]  gql.js
├── [2.7K]  printErrors.js
├── [4.4K]  query.js
└── [ 224]  util
    ├── [  96]  __tests__
    │   └── [3.2K]  request.test.js
    ├── [ 407]  defer.js
    ├── [1.6K]  fs.js
    ├── [1.0K]  request.js
    └── [9.7K]  ssh.js

9 directories, 26 files
```

**After:**

```
tree -h
.
├── [ 128]  __tests__
│   ├── [2.8K]  printErrors.test.js
│   └── [2.7K]  query.test.js
├── [1.7K]  cli.js
├── [ 352]  commands
│   ├── [ 192]  __tests__
│   │   ├── [1.8K]  customer.test.js
│   │   ├── [1.8K]  init.test.js
│   │   ├── [2.0K]  project.test.js
│   │   └── [1.8K]  projects.test.js
│   ├── [3.0K]  customer.js
│   ├── [  13]  index.js
│   ├── [3.4K]  init.js
│   ├── [1.6K]  login.js
│   ├── [1.6K]  logout.js
│   ├── [3.0K]  project.js
│   ├── [ 128]  projectCommands
│   │   ├── [  96]  __tests__
│   │   │   └── [6.3K]  project-create.test.js
│   │   └── [ 11K]  create.js
│   └── [1.9K]  projects.js
├── [ 128]  config
│   ├── [  96]  __tests__
│   │   └── [1.2K]  config.test.js
│   └── [1.3K]  index.js
├── [ 339]  gql.js
├── [2.1K]  printErrors.js
├── [2.4K]  query.js
└── [ 224]  util
    ├── [  96]  __tests__
    │   └── [1.6K]  request.test.js
    ├── [ 292]  defer.js
    ├── [ 863]  fs.js
    ├── [ 904]  request.js
    └── [3.9K]  ssh.js

9 directories, 26 files
```

**Comment by Schnitzel** [comment]
@karlhorky 
awesome work :) 

I changed some small things as we have them recently updated in https://github.com/amazeeio/lagoon/blob/master/services/drush-alias/web/aliases.drushrc.php.stub

Unfortunately creating a project doesn't work:

```
$ yarn execute project create
Project details for 'ci-drupal':
╔═══════════════════════╤════════════════════════════════════════════╗
║ Project Name          │ ci-drupal                                  ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Customer              │ ci-customer                                ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Git URL               │ ssh://git@192.168.99.1:2222/git/drupal.git ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Active Systems Deploy │ lagoon_openshiftBuildDeploy                ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Active Systems Remove │ lagoon_openshiftRemove                     ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Branches              │ true                                       ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Pull Requests         │ true                                       ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Openshift             │ ci-local                                   ║
╟───────────────────────┼────────────────────────────────────────────╢
║ Created               │ 2018-05-10 17:50:33                        ║
╚═══════════════════════╧════════════════════════════════════════════╝
```

it basically just prints the current project instead of creating one

**Comment by karlhorky** [comment]
@Schnitzel I've updated the execute script to pass along all arguments (using `$*`) and also support the `--identity` option (removing the extra `execute-nolocal` script)

Does this also solve your use case for `execute-nolocal`? If not, then I will re-instate the script.

**Pull request merged**